contracts/RolesRegistry/SftRolesRegistrySingleRole.sol

```diff
@@ -0,0 +1,249 @@
+            _grantRoleData.tokenAmount
+        );
+
+        _transferFrom(
+            _grantRoleData.grantor,
+            address(this),
+            _grantRoleData.tokenAddress,
+            _grantRoleData.tokenId,
+            _grantRoleData.tokenAmount
+        );
+    }
+
+    function revokeRoleFrom(uint256 _nonce, bytes32 _role, address _grantee) external override {
+        require(_role == UNIQUE_ROLE, 'SftRolesRegistry: invalid role');
+
+        RoleData memory _roleData = roleAssignments[_nonce];
+        require(_grantee == roleAssignments[_nonce].grantee, 'SftRolesRegistry: invalid grantee or nonce not used');
+        DepositInfo memory _depositInfo = deposits[_nonce];
+
+        address caller = _findCaller(_roleData, _depositInfo);
+        if (_roleData.expirationDate > block.timestamp && !_roleData.revocable) {
+            // if role is not expired and is not revocable, only the grantee can revoke it
+            require(caller == _roleData.grantee, 'SftRolesRegistry: nonce is not expired or is not revocable');
+        }
+
+        delete roleAssignments[_nonce];
+
+        emit RoleRevoked(
+            _nonce,
+            UNIQUE_ROLE,
+            _depositInfo.tokenAddress,
+            _depositInfo.tokenId,
+            _depositInfo.tokenAmount,
+            _depositInfo.grantor,
+            _roleData.grantee
+        );
+    }
+
+    function withdraw(
+        uint256 _nonce
+    ) public onlyOwnerOrApproved(deposits[_nonce].grantor, deposits[_nonce].tokenAddress) {
+        DepositInfo memory _depositInfo = deposits[_nonce];
+        require(
+            roleAssignments[_nonce].grantee == address(0) ||
+                roleAssignments[_nonce].expirationDate < block.timestamp ||
+                roleAssignments[_nonce].revocable,
+            'SftRolesRegistry: token has an active role'
+        );
+
+        delete deposits[_nonce];
+        delete roleAssignments[_nonce];
+
+        _transferFrom(
+            address(this),
+            _depositInfo.grantor,
+            _depositInfo.tokenAddress,
+            _depositInfo.tokenId,
+            _depositInfo.tokenAmount
+        );
+
+        emit Withdrew(
+            _nonce,
+            _depositInfo.grantor,
+            _depositInfo.tokenAddress,
+            _depositInfo.tokenId,
+            _depositInfo.tokenAmount
+        );
+    }
+
+    function setRoleApprovalForAll(address _tokenAddress, address _operator, bool _isApproved) external override {
+        tokenApprovals[msg.sender][_tokenAddress][_operator] = _isApproved;
+        emit RoleApprovalForAll(_tokenAddress, _operator, _isApproved);
+    }
+
+    /** View Functions **/
+
+    function roleData(uint256 _nonce, bytes32 _role, address _grantee) external view returns (RoleData memory) {
+        require(_role == UNIQUE_ROLE, 'SftRolesRegistry: invalid role');
+        require(_grantee == roleAssignments[_nonce].grantee, 'SftRolesRegistry: invalid grantee or nonce not used');
```

Comment:
These requires are used in multiple places (including `revokeRoleFrom`), perhaps turn them into modifiers? I would also change the role error message to `role not supported`, and the grantee error message to `grantee mismatch`